### ISSUE TITLE
Small change required to successfully use XTL with nvcc CUDA compiler

### ISSUE THREAD

**Issue by gridley** [xtensor-stack/xtl#183]
Hi,

I was hoping that since XTL works with any C++14 compiler, the compilation of __host only__ CUDA 10 code should work fine with XTL. Normally, I would not attempt this and instead find a replacement that is out-of-the-box compatible with CUDA. However, I need to write some CUDA as part of a large codebase which uses xtensor, [openmc](https://github.com/openmc-dev/openmc). It does seem that xtensor isn't far off from being able to compile host code with nvcc, I only obtain the following errors when compiling the most basic of xtensor programs (the first one on its readthedocs):

```c++
gavin@gpad:~/scratch$ nvcc xtensor_example.cu 
/usr/local/include/xtl/xsequence.hpp:136:87: error: expected primary-expression before ‘unsigned’
         struct sequence_forwarder_impl<R, A, void_t<decltype(std::declval<R>().resize(std::size_t()))>>
                                                                                       ^~~~~~~~
/usr/local/include/xtl/xvariant_impl.hpp: In function ‘constexpr decltype(auto) mpark::visit(Visitor&&, Vs&& ...)’:
/usr/local/include/xtl/xvariant_impl.hpp:2676:96: error: parameter packs not expanded with ‘...’:
     return (detail::all({!vs.valueless_by_exception()...})
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                     ^                                                                                                      
/usr/local/include/xtl/xvariant_impl.hpp:2676:96: note:         ‘vs’
/usr/local/include/xtensor/xstrided_view_base.hpp: In instantiation of ‘void xt::detail::strided_view_args<adj_strides_policy>::fill_args(const S&, ST&&, std::size_t, xt::layout_type, const V&) [with S = xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>; ST = const xt::svector<long int, 4, std::allocator<long int>, true>&; V = std::vector<mpark::variant<long int, xt::xrange_adaptor<xt::placeholders::xtuph, long int, long int>, xt::xrange_adaptor<long int, xt::placeholders::xtuph, long int>, xt::xrange_adaptor<long int, long int, xt::placeholders::xtuph>, xt::xrange_adaptor<long int, xt::placeholders::xtuph, xt::placeholders::xtuph>, xt::xrange_adaptor<xt::placeholders::xtuph, long int, xt::placeholders::xtuph>, xt::xrange_adaptor<xt::placeholders::xtuph, xt::placeholders::xtuph, long int>, xt::xrange_adaptor<long int, long int, long int>, xt::xrange_adaptor<xt::placeholders::xtuph, xt::placeholders::xtuph, xt::placeholders::xtuph>, xt::xrange<long int>, xt::xstepped_range<long int>, xt::xall_tag, xt::xellipsis_tag, xt::xnewaxis_tag> >; adj_strides_policy = xt::detail::no_adj_strides_policy; std::size_t = long unsigned int]’:
/usr/local/include/xtensor/xstrided_view.hpp:593:1:   required from ‘auto xt::strided_view(E&&, const xstrided_slice_vector&) [with E = const xt::xarray_container<xt::uvector<double, std::allocator<double> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&; xt::xstrided_slice_vector = std::vector<mpark::variant<long int, xt::xrange_adaptor<xt::placeholders::xtuph, long int, long int>, xt::xrange_adaptor<long int, xt::placeholders::xtuph, long int>, xt::xrange_adaptor<long int, long int, xt::placeholders::xtuph>, xt::xrange_adaptor<long int, xt::placeholders::xtuph, xt::placeholders::xtuph>, xt::xrange_adaptor<xt::placeholders::xtuph, long int, xt::placeholders::xtuph>, xt::xrange_adaptor<xt::placeholders::xtuph, xt::placeholders::xtuph, long int>, xt::xrange_adaptor<long int, long int, long int>, xt::xrange_adaptor<xt::placeholders::xtuph, xt::placeholders::xtuph, xt::placeholders::xtuph>, xt::xrange<long int>, xt::xstepped_range<long int>, xt::xall_tag, xt::xellipsis_tag, xt::xnewaxis_tag> >]’
/usr/local/include/xtensor/xio.hpp:260:31:   required from ‘void xt::detail::recurser_run(F&, const E&, xt::xstrided_slice_vector&, std::size_t) [with F = xt::detail::printer<xt::xarray_container<xt::uvector<double, std::allocator<double> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>, void>; E = xt::xarray_container<xt::uvector<double, std::allocator<double> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>; xt::xstrided_slice_vector = std::vector<mpark::variant<long int, xt::xrange_adaptor<xt::placeholders::xtuph, long int, long int>, xt::xrange_adaptor<long int, xt::placeholders::xtuph, long int>, xt::xrange_adaptor<long int, long int, xt::placeholders::xtuph>, xt::xrange_adaptor<long int, xt::placeholders::xtuph, xt::placeholders::xtuph>, xt::xrange_adaptor<xt::placeholders::xtuph, long int, xt::placeholders::xtuph>, xt::xrange_adaptor<xt::placeholders::xtuph, xt::placeholders::xtuph, long int>, xt::xrange_adaptor<long int, long int, long int>, xt::xrange_adaptor<xt::placeholders::xtuph, xt::placeholders::xtuph, xt::placeholders::xtuph>, xt::xrange<long int>, xt::xstepped_range<long int>, xt::xall_tag, xt::xellipsis_tag, xt::xnewaxis_tag> >; std::size_t = long unsigned int]’
/usr/local/include/xtensor/xio.hpp:763:21:   required from ‘std::ostream& xt::pretty_print(const xt::xexpression<D>&, std::ostream&) [with E = xt::xarray_container<xt::uvector<double, std::allocator<double> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>; std::ostream = std::basic_ostream<char>]’
/usr/local/include/xtensor/xio.hpp:776:20:   required from ‘std::ostream& xt::operator<<(std::ostream&, const xt::xexpression<D>&) [with E = xt::xarray_container<xt::uvector<double, std::allocator<double> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>; std::ostream = std::basic_ostream<char>]’
xtensor_example.cu:18:15:   required from here
/usr/local/include/xtensor/xstrided_view_base.hpp:897:6: error: ‘void info’ has incomplete type
                         auto info = xtl::visit(slice_getter, slices[i]);
```

It seems that a relatively small number of changes could make this work. While I do know some stuff about templates, my knowledge isn't good enough to identify the origin of these errors. I imagine there is some C++17 feature being used here that nvcc doesn't recognize. Since xtensor is supposed to work with C++14, and nvcc is supposed to work with C++14 as well, I thought perhaps this would merit raising an issue.

Any help would be MASSIVELY appreciated.

**Comment by JohanMabille** [comment]
Hi,

There is no C++17 feature here, only valid C++14, so this looks like a compiler issue ( which does not mean that we should not try to find a workaround). Which version of nvcc do you use?

**Comment by gridley** [comment]
Thanks Mr. Mabille. Suppose I’ll cook up an MWE and take the nvidia forums then.

**Comment by JohanMabille** [comment]
We had some surprises with the `void_t`trick on other compilers in the past, so I wouldn't be surprised that we are hitting some corner case here. Can you try the following example to see if you still have the first issue?

```cpp
#include <vector>
#include <array>

#include "xtl/xsquence.hpp"

int main(int argc, char* argv[])
{
    using array_type = std::array<double, 3>;
    using vector_type = std::vector<double>;
    
    vector_type vsrc = { 0., 1., 2. };
    auto adst = xtl::forward_sequence<array_type>(vsrc);
    std::cout << adst[0] << std::endl;

    array_type asrc = {0., 1., 2. };
    auto vdst = xtl::forward_sequence<vector_type>(asrc);
    std::cout << vdst[0] << std::endl;

    return 0;
}
```

**Comment by gridley** [comment]
Hm. This seems to give an error on each of the forward_sequences you declared there. Here are the errors for each:

```c++
gavin@gpad:~/scratch$ nvcc mabille_mwe.cu 
mabille_mwe.cu(14): error: cannot deduce the return type of function "xtl::forward_sequence<R,A>(A &&) [with R=std::array<double, 3UL>, A=std::vector<double, std::allocator<double>> &]"
/usr/local/include/xtl/xsequence.hpp(33): here; it has not been defined

mabille_mwe.cu(18): error: cannot deduce the return type of function "xtl::forward_sequence<R,A>(A &&) [with R=std::vector<double, std::allocator<double>>, A=std::array<double, 3UL> &]"
/usr/local/include/xtl/xsequence.hpp(33): here; it has not been defined

mabille_mwe.cu(21): error: cannot deduce the return type of function "xtl::forward_sequence<R,A>(A &&) [with R=std::array<double, 3UL>, A=std::array<double, 3UL> &]"
/usr/local/include/xtl/xsequence.hpp(33): here; it has not been defined

mabille_mwe.cu(22): error: cannot deduce the return type of function "xtl::forward_sequence<R,A>(A &&) [with R=std::vector<double, std::allocator<double>>, A=std::vector<double, std::allocator<double>> &]"
/usr/local/include/xtl/xsequence.hpp(33): here; it has not been defined

4 errors detected in the compilation of "/tmp/tmpxft_00000c37_00000000-8_mabille_mwe.cpp1.ii".

```

I'm not familiar enough with XTL to tell what auto should deduce here, unfortunately.

**Comment by gridley** [comment]
So, I can confirm that the void_t trick works on this compiler. Wrote this:

```c++
#include <iostream>
#include <type_traits>
using namespace std;

// check for a member called asdf
template <typename, typename = void>
struct has_asdf : std::false_type {};
template <typename T>
struct has_asdf<T, std::void_t<decltype(T::asdf)>> : std::true_type {};

// two structs that will be checked for a member called asdf:
struct Duck
{
  double asdf;
};
struct Chicken
{
  double not_asdf;
};

int main()
{
  bool duck_has_asdf = has_asdf<Duck>::value;
  bool chicken_has_asdf = has_asdf<Chicken>::value;
  cout << "duck: " << duck_has_asdf << endl;
  cout << "chicken: " << chicken_has_asdf << endl;
}
```

And get the expected result from both compilers:

```
gavin@gpad:~/scratch$ g++ void_t_trick.C 
gavin@gpad:~/scratch$ ./a.out 
duck: 1
chicken: 0
gavin@gpad:~/scratch$ cp void_t_trick.C void_t_trick.cu
gavin@gpad:~/scratch$ nvcc void_t_trick.cu 
gavin@gpad:~/scratch$ ./a.out 
duck: 1
chicken: 0
```

So... what exactly would cause this mechanism to fail, I wonder?

**Comment by JohanMabille** [comment]
It might be the combination with other things (declval and invoking methods for instance). Thanks to your test, we know that inner type detection works. Out of curiosity, can you test with `xtl::void_t` instead of `std::void_t`? I'm pretty sure it will work, but let's be systematic.

Let's fully specify the return type to see what you get:

```cpp
#include <vector>
#include <array>

#include "xtl/xsquence.hpp"

int main(int argc, char* argv[])
{
    using array_type = std::array<double, 3>;
    using vector_type = std::vector<double>;
    
    vector_type vsrc = { 0., 1., 2. };
    array_type adst = xtl::forward_sequence<array_type>(vsrc);
    std::cout << adst[0] << std::endl;

    array_type asrc = {0., 1., 2. };
    vector_type vdst = xtl::forward_sequence<vector_type>(asrc);
    std::cout << vdst[0] << std::endl;

    adst = xtl::forward_sequence<array_type>(asrc);
    vdst = xtl::forward_sequence<vector_type>(vsrc);
    std::cout << adst[0] << std::endl;
    std::cout << vdst[0] << std::endl;

    return 0;
}
```

**Comment by gridley** [comment]
Good point on checking void_t. So, yes, I can confirm that xmeta_utils.hpp and using xt::void_t does the trick as well.

OK, so, making the change you suggest, I still get the same errors as above regarding "it has not been defined". I suppose it wasn't the auto there but rather something deeper. I'm going to try my best to isolate this, and I'm happy to test any more things if you have ideas. Thanks again for the tips.

**Comment by JohanMabille** [comment]
Ok, let's try some simplification, can you comment lines 33 and 34 in xsequence.hpp and try again? These are just a forward declaration for readability.

**Comment by gridley** [comment]
So, yes, it does seem that this made a difference in the error message that's more helpful. Instead, the error becomes

```
mabille_mwe.cu(14): error: no instance of overloaded function "xtl::forward_sequence" matches the argument list
            argument types are: (vector_type)

mabille_mwe.cu(18): error: no instance of overloaded function "xtl::forward_sequence" matches the argument list
            argument types are: (array_type)

mabille_mwe.cu(21): error: no instance of overloaded function "xtl::forward_sequence" matches the argument list
            argument types are: (array_type)

mabille_mwe.cu(22): error: no instance of overloaded function "xtl::forward_sequence" matches the argument list
            argument types are: (vector_type)
```

I can't explain why this would happen. My thought for this type of error is that both definitions are being incorrectly SFINAE'd out, but I don't see where this might be happening.

So, something else to consider is that if I try to compile an empty program after including xsequence.hpp, I get this error message:

```
/usr/local/include/xtl/xsequence.hpp:136:87: error: expected primary-expression before ‘unsigned’
         struct sequence_forwarder_impl<R, A, void_t<decltype(std::declval<R>().resize(std::size_t()))>>
```

So I suppose figuring this out first would be good. Changing std::size_t() to std::declval<std::size_t>() seems to fix this. Maybe the nvidia compiler isn't correctly grouping the parentheses here or something?

**Comment by JohanMabille** [comment]
Nice, we are progressing! Good catch for the `std::declval<std::size_t>()`. The idea was to write something like `void_t<decltype(std::declval<R>().resize(0))>` but in that case any compiler would see `0` as `int` and would fail at detecting the existence of `resize(size_t)`. Thus the usage of `std::size_t()`. I think most of the compilers understand it as the litteral `0`. But strictly speaking, we should avoid the constructor and use `std::declval` instead.

Let's specify all the argument and hope the compiler will give us a more meaningful error message (or at least the list of overloads it considers and why they fail):

```cpp
#include <vector>
#include <array>

#include "xtl/xsquence.hpp"

int main(int argc, char* argv[])
{
    using array_type = std::array<double, 3>;
    using vector_type = std::vector<double>;
    
    vector_type vsrc = { 0., 1., 2. };
    array_type adst = xtl::forward_sequence<array_type, vector_type&>(vsrc);
    std::cout << adst[0] << std::endl;

    array_type asrc = {0., 1., 2. };
    vector_type vdst = xtl::forward_sequence<vector_type, array_type&>(asrc);
    std::cout << vdst[0] << std::endl;

    adst = xtl::forward_sequence<array_type, array_type&>(asrc);
    vdst = xtl::forward_sequence<vector_type, vector_type&>(vsrc);
    std::cout << adst[0] << std::endl;
    std::cout << vdst[0] << std::endl;

    return 0;
}
```

**Comment by gridley** [comment]
Well, fully specifying the template seems to correctly compile! I suppose the remove_reference stuff is messing up the overloaded functions in some way. I’ll be playing with this...

**Comment by JohanMabille** [comment]
Ah indeed, the `remove_reference` is here to prevent the compiler from deducing the type and to force the caller to specify all the arguments. It is tje same principle as the one used in the implementation of `std::forward`. I should have read the code mire carefully, sorry for missing that!

Would you like to open a PR to fix the two issues you found ? (commenting out the forward declaration and the use of `std::declval`)

**Comment by gridley** [comment]
Yes, I’d love to make a PR. However, I’d also like to figure out the trouble with xvariant_impl which appears at the beginning of this issue before doing so, if that’s OK.

I can confirm that the changes we just found fix the error in forward_sequence when trying to compile a basic xtensor program with CUDA though. Yay!

**Comment by gridley** [comment]
By the way, it seems that leaving the forward declaration in is fine. So it’s up to you whether to take it out.

**Comment by gridley** [comment]
So, it seems that the maintainer of the variant repository hasn't responded to a similar PR for a bit over a month, so could we maybe go ahead and merge this? The test system in master for variant is failing according to that [issue](https://github.com/mpark/variant/issues/70)/ [still unmerged PR](https://github.com/mpark/variant/pull/71).

**Comment by ax3l** [comment]
Congrats, your patch landed now in upstream: https://github.com/mpark/variant/commit/3c7fc8266bb46046b42c2dc2663f9f505f0cec28 :)

**Comment by gridley** [comment]
OK, so, when would be good to make the change from mpark/variant to here?

**Comment by JohanMabille** [comment]
Wow I had totally forgotten this, sorry! I can do it next week, or you can open a PR before if you prefer. We just need to check whether [xtensor](https://github.com/xtensor-stack/xtensor) and [xframe](https://github.com/xtensor-stack/xframe) build with this change to know if the release will be backward-compatible or not.

**Comment by gridley** [comment]
No worries! Well, I know that xtensor works with this change since I've run it locally many times. Just thought it would be good to get this change made.

**Comment by JohanMabille** [comment]
Good to know! So as you wish, if you open the PR I will merge it and release right after. Otherwise, I will do it on Monday.

**Comment by JohanMabille** [comment]
Fixed with #188
